### ISSUE TITLE
False positive for "unused-variable" with name not used as variable

### ISSUE THREAD

**Issue by mwchase** [PyCQA/pylint#3225]
### Steps to reproduce
1. Create the following directory structure and file contents:

`tree` output:
```
.
└── repro_dir
    ├── errors.py
    └── source.py
```
`source.py`:
```
"""Docstring."""


def cant_set_new_functions(function):
    """Docstring."""
    return function.__get__(None, None)
```
`errors.py`:
```
"""Docstring."""

import source


not source.cant_set_new_functions(object.__eq__)
```
2. Run `pylint repro_dir/`

### Current behavior
Reported errors:
```
************* Module errors
repro_dir/errors.py:6:0: W0106: Expression "not source.cant_set_new_functions(object.__eq__)" is assigned to nothing (expression-not-assigned)
************* Module source
repro_dir/source.py:1:0: W0612: Unused variable '__eq__' (unused-variable)
```
The bug requires that the attribute being accessed has double underscores, and it must actually exist. The type that it's accessed from doesn't matter. I'm pretty sure the following things cause the bug not to manifest:

- Defining the function in the same module in which it is called
- Removing the "not"
- Removing anything from the body of the function

### Expected behavior
```
************* Module errors
repro_dir/errors.py:6:0: W0106: Expression "not source.cant_set_new_functions(object.__eq__)" is assigned to nothing (expression-not-assigned)
```

### pylint --version output
```
pylint 2.4.3
astroid 2.3.2
Python 3.8.0 (default, Oct 16 2019, 15:47:19) 
[Clang 11.0.0 (clang-1100.0.33.8)]
```

**Comment by PCManticore** [comment]
Thank you for the clear reproduction examples, super useful. I can confirm the bug.

**Comment by PCManticore** [comment]
Congratulations, you won the bug of the year. 😅 This is fixed in astroid's `master`, the linked commit should explain more how this bug came to be. It requires both `__get__` and `object.__eq__` to reproduce.
